public/percent-complete/percent-complete.js

```diff
@@ -0,0 +1,265 @@
+import { JiraIssue } from "../shared/issue-data/issue-data.js";
+import { estimateExtraPoints } from "../confidence.js";
+import { millisecondsToDay, parseDate8601String } from "../date-helpers.js";
+
+/**
+ * @param { JiraIssue[] } issues
+ * @param { PercentCompleteOptions } options
+ */
+export function percentComplete(issues, options) {
+  /** @type {PercentCompleteOptions} */
+  const completeOptions = {
+    ...options,
+    toWholeDay: options.toWholeDay ?? Math.round,
+  };
+
+  const map = buildMap(issues, completeOptions);
+  setDurations(map, completeOptions);
+
+  console.log("percentComplete: map=", map);
+}
+
+/**
+ * @export
+ * @param {JiraIssue[]} issues
+ * @param {PercentCompleteOptions} options
+ * @returns {Record<string, CompletionIssue>>}
+ */
+export function buildMap(issues, options) {
+  return issues.reduce((acc, issue) => {
+    const key = options.getIssueKey(issue);
+    const parentKey = options.getParentKey(issue);
+
+    acc[key] = {
+      childKeys: acc[key]?.childKeys ?? [],
+      confidence: options.getConfidence(issue),
+      dueDate: options.getDueDate(issue),
+      hierarchyLevel: options.getHierarchyLevel(issue),
+      key,
+      parentKey,
+      startDate: options.getStartDate(issue),
+      storyPoints: options.getStoryPoints(issue),
+      storyPointsMedian: options.getStoryPointsMedian(issue),
+      type: options.getType(issue),
+    };
+
+    if (parentKey) {
+      if (acc[parentKey]) {
+        acc[parentKey].childKeys.push(key);
+      } else {
+        acc[parentKey] = { childKeys: [key] };
+      }
+    }
+
+    return acc;
+  }, {});
+}
+
+/**
+ * @export
+ * @param { Record<string, CompletionIssue> } map
+ * @param { PercentCompleteOptions } options
+ * @param { string[] } [parentTypeIssueKeys] Contents will be mutated to include duration
+ * information.
+ */
+export function setDurations(map, options, parentTypeIssueKeys = []) {
+  /**
+   * @param {string} key Sum the duration of the children (and children's children, etc.) for this
+   * issue.
+   * @return {number} Will be -1 if this function fails to generate child durations.
+   */
+  function sumIssueChildren(key) {
```

Comment:
Where does this infinitely recurse? I see it will get all children of children.  But I can't see the spot where it would go beyond that.